areal/experimental/openai/client.py

```diff
@@ -81,52 +81,73 @@ def _ensure_message_dict_list(
-            normalized.append(_item)
-        elif isinstance(item, BaseModel):
-            normalized.append(item.model_dump(exclude_none=True))
+        if isinstance(item, dict) or isinstance(item, BaseModel):
+            normalized.append(_normalize(item))
         else:
             raise TypeError(
                 f"{name}[{index}] must be a dict or a BaseModel; got {type(item).__name__}"
             )
     return normalized
 
 
+def get_extra_tokens_after_im_end(tokenizer: "PreTrainedTokenizerFast") -> list[int]:
+    if not hasattr(get_extra_tokens_after_im_end, "_cache"):
+        im_end_token_id = tokenizer.eos_token_id
+        simple_chat_tokens = tokenizer.apply_chat_template(
+            [{"role": "user", "content": ""}],
+            tokenize=True,
+            add_generation_prompt=False,
+        )
+        im_end_index = (
+            len(simple_chat_tokens)
+            - 1
+            - simple_chat_tokens[::-1].index(im_end_token_id)
+        )
+        get_extra_tokens_after_im_end._cache = simple_chat_tokens[im_end_index + 1 :]
+    return get_extra_tokens_after_im_end._cache
+
+
 def concat_prompt_token_ids_with_parent(
     message_list: list[dict],
     parent: InteractionWithTokenLogpReward | None,
     tokenizer: "PreTrainedTokenizerFast",
-    start: str,
-    end: str,
+    tools: Iterable[ChatCompletionToolParam] | None = None,
+    extra_body: Body = {},
 ) -> list[int]:
     """Concatenate prompt token IDs with parent interaction's tokens."""
     parent_tokens = []
     if parent is not None:
         if parent.model_response is None:
             raise ValueError("Parent interaction has no model_response.")
+        # TODO: (yulangz) how to handle here when stop is set?
         parent_tokens = (
-            parent.model_response.input_tokens + parent.model_response.output_tokens
+            parent.model_response.input_tokens
+            + parent.model_response.output_tokens  # with stop tokens
         )
     # By default, follows Qwen3 chat template.
-    message_strs = []
-    for msg in message_list:
-        message_strs.append(f"{start}{msg['role']}\n{msg['content']}{end}\n")
-        warn = msg["role"] not in ["user", "system"] or any(
-            k not in ["content", "role"] for k in msg.keys()
-        )
-        if warn:
-            logger.warning(
-                "When using 'concat' chat template, only 'user' or 'system' role "
-                "messages with 'content' field are properly handled. Other roles "
-                "or extra fields may lead to unexpected tokenization results. "
-                "Please ensure user-side messages are only of 'user' or 'system' "
-                "role with 'content' field."
-            )
-    message_strs.append(f"{start}assistant\n")
-    prompt_token_ids = parent_tokens + tokenizer.encode("".join(message_strs))
+    child_tokens = tokenizer.apply_chat_template(
+        message_list,
+        tools=tools,
```

Comment:
Actually, tools are applied within the system prompt. Do we need to apply this tool for each round here?